CefSharp.Wpf/Internals/MouseAdjustor.cs

```diff
@@ -0,0 +1,132 @@
+        private int xOffset;
+
+        /// <summary>
+        /// The y-offset.
+        /// </summary>
+        private int yOffset;
+
+        /// <summary>
+        /// The original-rect.
+        /// </summary>
+        private Rect originalRect;
+
+        /// <summary>
+        /// The teleporting-rect.
+        /// </summary>
+        private Rect teleportingRect;
+
+        public System.Windows.Point Update(Rect originalRect, Rect viewRect)
+        {
+            int x = originalRect.X,
+                prevX = originalRect.X,
+                y = originalRect.Y,
+                prevY = originalRect.Y,
+                xOffset = 0,
+                yOffset = 0;
+
+            // If popup goes outside the view, try to reposition origin
+            if (originalRect.X + originalRect.Width > viewRect.Width)
+            {
+                x = viewRect.Width - originalRect.Width;
+                xOffset = prevX - x;
+            }
+
+            if (originalRect.Y + originalRect.Height > viewRect.Height)
+            {
+                y = y - originalRect.Height - 20;
+                yOffset = prevY - y;
+            }
+
+            // If x or y became negative, move them to 0 again
+            if (x < 0)
+            {
+                x = 0;
+                xOffset = prevX;
+            }
+
+            if (y < 0)
+            {
+                y = 0;
+                yOffset = prevY;
+            }
+
+            if (x != prevX || y != prevY)
+            {
+                this.xOffset = xOffset;
+                this.yOffset = yOffset;
+
+                Rect teleportingRect = new Rect(x, y, x + originalRect.Width, y + originalRect.Height);
+
+                this.originalRect = originalRect;
+                this.teleportingRect = teleportingRect;
+
+                if (this.originalRect.Y < this.teleportingRect.Y + this.teleportingRect.Height)
+                {
+                    var newY = this.teleportingRect.Y + this.teleportingRect.Height;
+                    this.originalRect = new Rect(originalRect.X, newY, originalRect.Width, originalRect.Y + originalRect.Height - newY);
+                }
+            }
+
+            return new System.Windows.Point(x, y);
+        }
+
+        /// <summary>
+        /// Resets the offsets and original-rect.
+        /// </summary>
+        public void Reset()
+        {
+            this.xOffset = 0;
+            this.yOffset = 0;
+
+            this.originalRect = new Rect();
+            this.originalRect = new Rect();
+        }
+
+        /// <summary>
+        /// Adjusts the mouse-coordinates when the popup is visible.
+        /// </summary>
+        /// <param name="point">The original point.</param>
+        /// <returns>The adjusted point if needed, else the original point.</returns>
+        public Point GetAdjustedMouseCoords(System.Windows.Point point)
+        {
+            return !this.IsInsideOriginalRect(point) && IsInsideTeleportingRect(point)
```

Comment:
We shouldn't need to call these methods when there is no popup currently active or when then the popup is active and it it's shown in it's default location.

We should track this with a simple boolean property for performance reasons.

Reply:
Thank you again! This is the only comment where I am unsure about the implementation. Unfortunately, I have no idea how to get a bool for "isOpen" from the stack above. Do you have an idea?

Reply:
If you add a class level variable (at private field of type bool), you can keep track of if the popup is open and the position was transformed. 

if the bool is false simply return the point, if it's true then perform the checks.

- If popup is closed, then bool will be false
- If popup is open, yet not transformed then bool will be false
- If popup is open and transformed then bool will be true.

Reply:
Thank you. I added a commit to the PR with the changes.